### DISCUSSION TITLE
Midi clock?

### DISCUSSION THREAD

**Discussion by CharHeap** [thetwom/toc2#83] [Ideas]
Hey, first of all: really loving the app (+ your tuner app), thanks!

I hope this is the right place to suggest this feature/ask for help.

For my band we need a **midi clock signal** to e.g. sync a delay pedal with the current bpm automatically.

I was able to add all the required midi device code, open a port and send the clock pulse byte on every quarter note.
However, the clock pulse has to be sent **24 times for every quarter note**.

My quick and dirty workaround was creating a scene with 1 loud and 23 muted quarter notes and multiplying the quarter bpm value by 24 (in the code). Of course the visuals were now broken (as they were triggered 24 times as fast) but this "works" for now.

I am not sure if such a midi clock feature would be useful to the average user (since this also requires a midi usb cable most people don't have lying around), so I am not really expecting you to add this feature. It would most likely just clutter the otherwise minimalistic app unnecessarily?

I would however be grateful for any help or hints on how I would implement this **properly**. I haven't dug through all of the code yet and am therefore not sure where I could create an additional callback which then gets triggered more often than the audio/visual events. Maybe there is a simple solution?

**Comment by thetwom** [comment]
Hi, thanks for the interest in the app. Good to hear, that you managed to get it working even though the callback was not called often enough. 

Currently, the callbacks are triggered in the part of the code, where the sound stream is generated. Without having double-checked  it, I would think, that the best way is to have an additional callback, which handles the trigger for midi signals.  I will try to have  a look within the next days, and at least give some hints, where this could be added. 

It is definitely a good question, whether it is worth to have  such a feature in the main version. From perspective of user interface, it seems to me, as it would need a new option in the settings, to enable this. Not sure, how much it increases the size of the binary. 
 
Do you know, where this factor of 24 comes from?

Maybe you could also share your code, I would be interested to see, how this midi trigger works.

**Reply by CharHeap** [reply]
I don't know where the 24 comes from, but these seem to be the _pulses per quarter note_ (PPQN) for most devices/midi clocks? Looks like depending on the use case other PPQN are used aswell, so maybe this should be configurable too.

Happy to share the code, hope it's fine if I just paste the most relevant part here.

Following the `enableVibration`/`disableVibration` functions in PlayerService.kt I added:
```kotlin
     /** Enable midi.
     * @param delayInMillis Initial delay in millis of a midi send in comparison to the sound.
     */
    private fun enableMIDI(delayInMillis: Int) {
        if (noteStartedHandler4MIDI != null)
            return
        
        val midiManager = getSystemService(MIDI_SERVICE) as MidiManager
        val midiDevices = midiManager.devices

        if (midiDevices.isNotEmpty()) {
            // TODO Always using first device for now..
            val deviceInfo = midiDevices[0]

            midiManager.openDevice(
                deviceInfo,
                { device ->
                    if (device != null) {
                        openedMIDIDevice = device

                        // TODO Always using port 0..
                        port = openedMIDIDevice?.openInputPort(0)
                    }
                }, null)
        }

        noteStartedHandler4MIDI = audioMixer?.createAndRegisterNoteStartedHandler(
            delayInMillis, NoteStartedHandler.CallbackWhen.NoteStarted, null
        ) { noteListItem, _, durationNanos, _, _ ->
            if (noteListItem != null) {
                // Midi port established
                if (port != null) {
                    val midiStartMessage = byteArrayOf(0xFA.toByte())
                    val midiClockMessage = byteArrayOf(0xF8.toByte())
                    // val midiStopMessage = byteArrayOf(0xFC.toByte())

                    // TODO Not sure if sending this "start byte" is even required
                    if(!sentStart) {
                        port?.send(
                            midiStartMessage, 0, midiStartMessage.size
                        )
                        sentStart = true
                    }

                    // Send clock pulse
                    port?.send(midiClockMessage, 0, midiClockMessage.size)
                }
            }
        }
    }
```
The logic probably isn't perfect, e.g. the midi device has to be available when the preference is toggled, otherwise the registered handler remains useless.. But you get the idea.

The `disableMidi` function just unregisters the handler, calls `close()` on both the port and device and sets them to `null`. It also sets `sentStart` to false, but not sure whether this initial start byte is even needed. I also don't use the stop byte, didn't seem to be necessary for our device.

There are also some declarations at the top for the new vars and minor changes in e.g. `onSharedPreferenceChanged` to call these functions, identical to how it's done for the vibration preference. Let me know if you want the complete PlayerService.kt file.

P.S.: I also added to AndroidManifest.xml:
```xml
    <uses-permission android:name="android.permission.MIDI" />
    <!-- are the following required? -->
    <uses-feature android:name="android.hardware.usb.host" android:required="true" />
    <uses-feature android:name="android.software.midi" android:required="true"/>
```

**Comment by thetwom** [comment]
Thanks for sharing the info and sharing your code. 

I had a look, whether there is a easy way to implement the 24 callbacks per quarter note. Unfortunately it seems not super-easy, but I have a second look :-). Problem is to keep everything nicely in sync with the actually played notes, especially while changing the speed. But there are ways, question is how easy it can be implemented. 

As for bringing changes to the main version, a big uncertainty is the need for extra permissions. I think it is good to keep the required permissions at the minimum. But sometimes features just need them. I don't say, this would be a blocker, but it would need a second thought :-).

**Comment by thetwom** [comment]
So ... in case you are still interested, I did now publish a successor app: [https://codeberg.org/thetwom/metronome](https://codeberg.org/thetwom/metronome). Here, I did at least allow to use callbacks based on subbeats, so exactly the basis, what you need. A lot changed there, but in case you are interested you could try to get midi working there in a much more consistent way ... and maybe at some time mainline it ...

**Comment by CharHeap** [comment]
Thank you very much 👍
